### PR TITLE
Fix secondary code freq

### PR DESCRIPTION
**Pull request by yookoala**
* fix code frequency issue as mentioned in #104.
* add previously missing artifact `compile` to .gitignore.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,7 @@ config.log
 config.status
 config.sub
 configure
+compile
```
Comment (bochecha) [inline .gitignore]: Hurray for the autotools periodically adding new files.

If someone wants to move us to Meson, I'd be very happy. 🙂

**Comment by bochecha** [comment]
Sorry for the time it took me to get to this, life has been a bit busy lately.

I see you decided to skip the simple fix for #77 entirely, and go straight for the more general fix. That's ok, it just makes it a bit harder to review. (it has been a very long time since I last looked at the `dbbuilder` code 😮)

The code looks ok at a first glance, I'll try to make some time to test it a bit before merging.

My main reservation on this change is that it has a big potential to make a lot of changes to the resulting database, and I personally don't (and in fact can't) know whether those changes are correct.

I'd like some reassurance from you that you did look at all the impacted lines, and ensured that for each of them, the code we list second should appear last.

[Your comment](https://github.com/Cangjians/ibus-cangjie/issues/77#issuecomment-269979778) lists 254 such lines. Are you positive that for each of them, the mapping we list second (or third, or ...) should appear last in a list of candidates?

That same comment says:

> After removing the x-disambiguation mappings

With this `dbbuilder` change, what happens to the lines with x-disambiguation mappings?

Finally, if this `dbbuilder` change covers the special-case we already had for `氐`, should we remove that special case from `table.txt`?

**Comment by yookoala** [comment]
I made sure that the resulting changes fixed the order of quick and cangjie candidates.

Besides, extra mappings were there as an extra efforts to better mapping. They do not have place in Windows or Mac implementation ,hence do not have any expected order to preserve in the first place.

**Comment by yookoala** [comment]
Pressed wrong button.

**Comment by yookoala** [comment]
I will give a thought on x-disambiguation cases. Reply later.

**Comment by bochecha** [comment]
> I will give a thought on x-disambiguation cases. Reply later.

Do note that I'm not asking you to fix them in this pull request.

All I'm asking for is that this pull request doesn't make them worse than they currently are. If you tell me this pull request doesn't change anything for them, then that's just fine.

**Comment by bochecha** [comment]
>> I will give a thought on x-disambiguation cases. Reply later.

> Do note that I'm not asking you to fix them in this pull request.
> 
> All I'm asking for is that this pull request doesn't make them worse than they currently are. If you tell me this pull request doesn't change anything for them, then that's just fine.

So I thought some more on this, and ran some scripts to analyse things a bit.

My conclusion is that your change is completely orthogonal to the X-disambiguation issue.

However, playing a bit with some Python script to analyse things (only for Cangjie5), I found that there are currently 50 issues with badly-ordered results due to poor handling of the X-disambiguation. But with your change, that seems to fall down to less than 18.

So even though your change has no relationship whatsoever to the X-disambiguation issue, it seems to make things slightly better, by chance. :slightly_smiling_face: 

If you can have a look into this to make sure I'm not completely wrong, that would be great.

**Comment by bochecha** [comment]
I've been staring at our data for some a few days, and I think I found a problem with this pull request.

Your change fixes things like the ordering for `竹竹` (`h*h`) because:

1. `秀` has codes `竹木弓竹尸` (`hdnhs`) and `竹木弓尸竹` (`hdnsh`);
2. the second code is an "extra" mapping, not the canonical one;
3. you make the frequency apply only to the canonical mapping, not the extra one;

As a result, the extra mapping is still available, but it won't interfere with what people expect.

Now take another example: `迉`, which has codes `難卜尸` (`xys`) and `卜尸` (`ys`).

With your change, the frequency is kept only for the first code, which is x-disambiguated. It seems to me that the `ys` code should also have the frequency associated to it. Is that correct? Can you check where Windows places this character when typing `ys`?

**Comment by yookoala** [comment]
I agree there is problem with x-disambiguated code. And I am not sure how we should properly deal with them.

One solution I can think of is to have them in different line than the canonical code; while another is to have all the extra codes in separate line with 0 as frequency (and undo my `dbbuilder` changes). Both could practical solve the issue.

What I am thinking about now is "should we have the code table compiled in a more accessible format, like yml, then compile back to our table txt format. The problem with the current data is to contain too many implied rules and reasoning. An ideal format is to express properly each coding, the source and the reason behind that (i.e. canonical codes, extra codes, canonical position, frequency, x-disambiguated). If we have a better raw data format, perhaps we can reason with how we compile with table.txt later. It would also be great for other developer to use the data as they please.

I would also need to see if x-disambiguated code have official mappings. If not, maybe we should simply compile them from the mapping data instead of hard coding.

I'll do a bit studying and then refresh this PR this week. Stay tune.

Thanks.

**Comment by bochecha** [comment]
> If we have a better raw data format, perhaps we can reason with how we compile with table.txt later

That's what I'm working on right now.

The problem is that while doing this, I'm finding issues in our data, like the `覇` I told you on Matrix, or this here with x-disambiguated codes.

--- 

I think what I want for x-disambiguated code is that the xcode has the same frequency as the corresponding code.

For example, in our table if a character has codes abc,xabc,def,xdef and frequency 1000, then:

* codes abc and xabc would keep their frequency 1000;
* codes def and xdef would be given the frequency 0;

This is essentially what you're doing in this pull request, but you're giving the frequency only to "the first code". What I'm suggesting is giving it to "the first couple of code/xcode".

Does that make sense?

(all in all, this problem is one more symptom of an incorrect decision we took a long time ago: associating the "frequency" to the character, when it really should have been associated to the code)

**Comment by bochecha** [comment]
> another is to have all the extra codes in separate line with 0 as frequency (and undo my dbbuilder changes).

I think that's what I want.

This pull request basically encodes some data (the frequency for secondary codes) in the `dbbuilder` tool. The data should all be in one place, and the tools manipulating that data should have no knowledge encoded into them.

**Comment by bochecha** [comment]
I have a branch ready which should fix the data while avoiding the issues we found in this approach:

* only the source table is modified, the `dbbuilder` isn't;
* x-disambiguated codes stay grouped with their "canonical" code;

I'll send it as a pull request once we merge (or reject) #108. :slightly_smiling_face:

**Comment by bochecha** [comment]
We have moved to Freedesktop: https://gitlab.freedesktop.org/cangjie/libcangjie

Please continue discussion there.

**Pull request closed without merging**